---
title: "hematopoietic stem cells"
author: "Cindy Pino-Barrios Gary Chan"
date: "2023-05-05"
output: html_document
---
# Introduction

Constructed cell trajectories with Monocle 3 using single-cell ATAC-seq data. 


## Data Info

We used single-cell ATAC-seq data set containing human CD34+ hematopoietic stem and progenitor cells published by https://www.nature.com/articles/s41587-019-0206-z

The processed data is available on NCBI GEO here: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE129785

The vignette:https://stuartlab.org/signac/articles/monocle.html


## How did we convert data between Seurat (used by Signac) and CellDataSet (used by Monocle 3) formats, weused a conversion function in the SeuratWrapper package. 


```{r install}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EnsDb.Hsapiens.v75")


BiocManager::install("biovizBase") # this is needed 

```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Loaded the packages - Monocle 3 was challenging to load predominantly because some of the packages are not available and the documentation is not the best. The most difficult challenge was installing 'leidenbase'. 

```{r cars}
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(monocle3)
library(Matrix)
library(ggplot2)
library(patchwork)
library(dittoSeq)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v75)
set.seed(1234)
```


# Data Loading

```{r}

filepath <- "/Users/cindybarrios/Desktop/SpringQuarter/BMI_210-Immunology/project_bmi210/hematopoietic_stem-scATAC/GSE129785_scATAC-Hematopoiesis-CD34/"
```

***scATAC-Hematopoiesis-CD34 contains three files:***

  - peaks
  - cell_barcodes
  - mtx (matrix) 


***Progenitor data consisted of one file:***

  - GSM3722029_CD34_Progenitors_Rep1_fragments.tsv.gz

  However, you only download the tsv.gz, but you also need  tsv.gz.tbi which includes the index for it   to read the file. I used samtools tabix function to generate a new tsv.gz.tbi from tsv.gz. 
  Note: The tsv.gz.tbi file needs to be in the same directory as the tsv.gz
  - ***used tabix, for example: tabix -p bed ***


```{r accessing the data}

peaks <- read.table(paste0(filepath, "GSE129785_scATAC-Hematopoiesis-CD34.peaks.txt.gz"), header = TRUE)
cells <- read.table(paste0(filepath, "GSE129785_scATAC-Hematopoiesis-CD34.cell_barcodes.txt.gz"), header = TRUE, stringsAsFactors = FALSE)
rownames(cells) <- make.unique(cells$Barcodes)

mtx <- readMM(file = paste0(filepath, "GSE129785_scATAC-Hematopoiesis-CD34.mtx.gz"))
mtx <- as(object = mtx, Class = "CsparseMatrix") 

```

```{r}
colnames(mtx) <- rownames(cells)
rownames(mtx) <- peaks$Feature
```

```{r}
head(mtx)
```

```{r}
prog_frag = "/Users/cindybarrios/Desktop/SpringQuarter/BMI_210-Immunology/project_bmi210/hematopoietic_stem-scATAC/GSE129785_scATAC-Hematopoiesis-CD34/GSM3722029_CD34_Progenitors_Rep1_fragments.tsv.gz"
```

 - Read the count matrix and progenitor data using CreateChromatinAssay from Signac
    - load the matrix and progenitor 
 - CreateSeuratObject using the chromatin assay we just made setting it as ATAC and combining it with meta.data cells 


```{r reading the files}
bone_assay <- CreateChromatinAssay(
  counts = mtx,
  min.cells = 5,
  fragments = prog_frag,
  sep = c("_", "_"),   # the mtx files is separated by _ and _ 
  genome = "hg19" # specify the genome - human hg19 
)
bone <- CreateSeuratObject(
  counts = bone_assay,
  meta.data = cells,
  assay = "ATAC"
)
```

Let's take a look at our data:
```{r}
bone
bone[[]]
```
We have a Seurat object with 9873 cells with 570650 features 
If we look at the data, we can see some of the meta data that the 10x pipeline calculated 

```{r subset to only. include CD34 progenitor cells}
# The dataset contains multiple cell types
# We can subset to include just one replicate of CD34+ progenitor cells
bone <- bone[, bone$Group_Barcode == "CD34_Progenitors_Rep1"]

# add cell type annotations from the original paper
cluster_names <- c("HSC",   "MEP",  "CMP-BMP",  "LMPP", "CLP",  "Pro-B",    "Pre-B",    "GMP",
                  "MDP",    "pDC",  "cDC",  "Monocyte-1",   "Monocyte-2",   "Naive-B",  "Memory-B",
                  "Plasma-cell",    "Basophil", "Immature-NK",  "Mature-NK1",   "Mature-NK2",   "Naive-CD4-T1",
                  "Naive-CD4-T2",   "Naive-Treg",   "Memory-CD4-T", "Treg", "Naive-CD8-T1", "Naive-CD8-T2",
                  "Naive-CD8-T3",   "Central-memory-CD8-T", "Effector-memory-CD8-T",    "Gamma delta T")
num.labels <- length(cluster_names)
names(cluster_names) <- paste0( rep("Cluster", num.labels), seq(num.labels) )
bone$celltype <- cluster_names[as.character(bone$Clusters)]

bone[["ATAC"]]

```



Next we can add gene annotations for the hg19 genome to the object. This will be useful for computing quality control metrics (TSS enrichment score) and plotting.

***We need to do this before pre-processing to determine where the peaks fall relative to where the genes are***
```{r gene annotations}

# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75) # get the range from diff genes from the human database

# change to UCSC style since the data was mapped to hg19
seqlevelsStyle(annotations) <- 'UCSC'

# add the gene information to the object
Annotation(bone) <- annotations # add the annotations to our data 
```

```{r}
head(annotations) # take a look at how this annotations look like - we have ENSMB the location and the gene name 
```


# Quality Control 

Weâ€™ll compute TSS enrichment, nucleosome signal score, and the percentage of counts in genomic blacklist regions for each cell, and use these metrics to help remove low quality cells from the datasets.

  - ***Calculate NucleosomeSignal***: which calculates the ratio fragments that are between 147 - 294 base pairs and less than 147 because the average rate to wrap around one histone is approximately 147 base pairs. This way we can calculate the ratio of fragments from mononucleosomes vs fragments that are nucleosome free. This adds a new column in our meta-data. 
  - ***Calculate TSSEnrichment***: using our annotations, we can get the transcription start sites (TSS) based on the gene location
  - ***BlackList_fraction*** : we have a column called blacklist_region_fragments, the number of fragments that map into black list regions *aka: junk regions o regions that are known to be no good. The blacklist_region_fragments divided by the peak_region_fragments.

*This steps takes a few mins to process*

```{r}
bone <- TSSEnrichment(bone)
bone <- NucleosomeSignal(bone)
bone$blacklist_fraction <- FractionCountsInRegion(bone, regions = blacklist_hg19)
```

We can now view all the things we need to filter out from the data. 
```{r}

VlnPlot(
  object = bone,
  features = c("nCount_ATAC", "TSS.enrichment", "nucleosome_signal", "blacklist_fraction"),
  pt.size = 0.1,
  ncol = 4
)
```
```{r}
bone <- bone[, (bone$nCount_ATAC < 50000)] # very simple filtering!!! I could go back and filter more later
```



# Data Preprocessing

   - Normalized the data using TFIDF 
   - Find top features 
   - Run SVD - singular value decomposition (analogous to PCA)
   - DepthCor - plots the correlation between depth and reduced dimensional components
      - we are not using the first component because it is an artifact

```{r}
bone <- RunTFIDF(bone)
bone <- FindTopFeatures(bone, min.cells = 10)
bone <- RunSVD(bone, n = 100)
DepthCor(bone)
```
  - RunUMAP
  - FindNeighbors
  - FindClusters (algorithm 3) = 

*We are using 2 to 50 dimensions*

```{r}

bone <- RunUMAP(
  bone,
  reduction = "lsi",
  dims = 2:50,
  reduction.name = "UMAP"
)
```


```{r}

bone <- FindNeighbors(bone, dims = 2:50, reduction = "lsi")
bone <- FindClusters(bone, resolution = 0.8, algorithm = 3)
```


```{r}
dittoDimPlot(bone, var = 'Clusters', reduction.use = 'UMAP',labels.repel = T,
             do.label = T,
             legend.show = F)
```


```{r}

DimPlot(bone, label = TRUE) + NoLegend()
```

Assign each cluster to the most common cell type based on the original annotations from the paper

```{r}
for(i in levels(bone)) {
  cells_to_reid <- WhichCells(bone, idents = i)
  newid <- names(sort(table(bone$celltype[cells_to_reid]),decreasing=TRUE))[1]
  Idents(bone, cells = cells_to_reid) <- newid
}
bone$assigned_celltype <- Idents(bone)
```


```{r}
dittoDimPlot(bone, var = "assigned_celltype", reduction.use = 'UMAP',labels.repel = T,
             do.label = T,
             legend.show = T)
```
Next we can subset the different lineages and create a trajectory for each lineage. another way to build a trajectory is to use the whole dataset and build separate pseudotimes trajectories for the different cell partitions found by monocle3


```{r}

DefaultAssay(bone) <- "ATAC"

erythroid <- bone[,  bone$assigned_celltype %in% c("HSC", "MEP", "CMP-BMP")]
lymphoid <- bone[, bone$assigned_celltype %in% c("HSC", "LMPP", "GMP", "CLP", "Pro-B", "pDC", "MDP", "GMP")]
```

```{r}
erythroid@meta.data$assigned_celltype
```


```{r}
dittoDimPlot(erythroid, var = 'assigned_celltype', reduction.use = 'UMAP')
dittoDimPlot(lymphoid, var = 'assigned_celltype', reduction.use = 'UMAP')
```


## Building trajectories with monocle3

we can convert the Seurat object to celldataset object using the cell-data-set() function from the seuratwrappers and build the trajectories using monocle3. we;ll do this separately for erythoid and lymphoid lineage, but you could explaore other strategies building a trajectory for all lineages together. 
***if you don't know the cell types we could use graph_test here***
Then, You can explore the genes in each module or conduct gene ontology enrichment analysis on them to glean insights about which cell types are present. Suppose after doing this we have a good idea of what the cell types in the partition are. Let's recluster the cells at finer resolution and then see how they overlap with the clusters in the partition:

```{r}
erythroid.cds <- as.cell_data_set(erythroid) # Converst Seurat Object to celldataset using SeuratWrapper
erythroid.cds <- cluster_cells(cds = erythroid.cds, reduction_method = "UMAP")
erythroid.cds <- learn_graph(erythroid.cds, use_partition = TRUE)

lymphoid.cds <- as.cell_data_set(lymphoid)
lymphoid.cds <- cluster_cells(cds = lymphoid.cds, reduction_method = "UMAP")
lymphoid.cds <- learn_graph(lymphoid.cds, use_partition = TRUE)
```



To compute pseudotime estimates for each trajectory we need to decide what the start of each trajectory is. in our case, we know that the hematopoietic stem cells are the progenitors of to her cell types in the trajectory, so er can set these cells as the root of the trajectory. Monocle3 includes an interactive function to select cells as the root nodes int he graph. This function will be launched if calling order-Cells() without specifying the row_cells parameter. Here we pre selected some cells as the root, and saved these to a file for reproducibility. 

```{r}

# load the pre-selected HSCs
hsc <- readLines("/Users/cindybarrios/Desktop/vignette_data/GSE129785/GSE129785_scATAC-Hematopoiesis-CD34/hsc_cells.txt")

```

```{r}

# order cells
erythroid.cds <- order_cells(erythroid.cds, reduction_method = "UMAP", root_cells = hsc)
lymphoid.cds <- order_cells(lymphoid.cds, reduction_method = "UMAP", root_cells = hsc)

# plot trajectories colored by pseudotime
plot_cells(
  cds = erythroid.cds,
  color_cells_by = "pseudotime",
  show_trajectory_graph = TRUE
)
```



```{r}
plot_cells(
  cds = lymphoid.cds,
  color_cells_by = "pseudotime",
  show_trajectory_graph = TRUE
)
```

```{r}

bone <- AddMetaData(
  object = bone,
  metadata = erythroid.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Erythroid"
)

bone <- AddMetaData(
  object = bone,
  metadata = lymphoid.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Lymphoid"
)
```


```{r}
FeaturePlot(bone, c("Erythroid", "Lymphoid"), pt.size = 0.1) & scale_color_viridis_c()
```


Thanks to the developers of Monocle 3, especially Cole Trapnell, Hannah Pliner, and members of the Trapnell lab. If you use Monocle please cite the Monocle papers.

